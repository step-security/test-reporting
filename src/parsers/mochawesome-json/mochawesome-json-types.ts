export interface MochawesomeJson {
  stats: MochawesomeJsonStat
  results: MochawesomeJsonSuite[]
}

export interface MochawesomeJsonStat {
  duration: number
}

export interface MochawesomeJsonSuite {
  tests: MochawesomeJsonTest[]
  suites: MochawesomeJsonSuite[]
  fullFile: string
  title: string
}

export interface MochawesomeJsonTest {
  title: string
  fullTitle: string
  duration: number
  pass: boolean
  fail: boolean
  skipped: boolean
  err: MochawesomeJsonTestError
}

export interface MochawesomeJsonTestError {
  message: string
  estack: string
  diff: string | null // TODO - Not sure if string
}
