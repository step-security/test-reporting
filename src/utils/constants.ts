export enum Outputs {
  runHtmlUrl = 'runHtmlUrl'
}
