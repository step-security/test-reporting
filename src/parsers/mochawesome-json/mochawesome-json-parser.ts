import {ParseOptions, TestParser} from '../../test-parser'
import {
  TestCaseError,
  TestCaseResult,
  TestExecutionResult,
  TestGroupResult,
  TestRunResult,
  TestSuiteResult
} from '../../test-results'
import {getExceptionSource} from '../../utils/node-utils'
import {getBasePath, normalizeFilePath} from '../../utils/path-utils'
import {MochawesomeJson, MochawesomeJsonSuite, MochawesomeJsonTest} from './mochawesome-json-types'

export class MochawesomeJsonParser implements TestParser {
  assumedWorkDir: string | undefined

  constructor(readonly options: ParseOptions) {}

  async parse(path: string, content: string): Promise<TestRunResult> {
    const mochawesome = this.getMochawesomeJson(path, content)
    const result = this.getTestRunResult(path, mochawesome)
    result.sort(true)
    return Promise.resolve(result)
  }

  private getMochawesomeJson(path: string, content: string): MochawesomeJson {
    try {
      return JSON.parse(content)
    } catch (e) {
      throw new Error(`Invalid JSON at ${path}\n\n${e}`)
    }
  }

  private getTestRunResult(resultsPath: string, mochawesome: MochawesomeJson): TestRunResult {
    const suitesMap: {[path: string]: TestSuiteResult} = {}

    const results = mochawesome.results

    const getSuite = (fullFile: string): TestSuiteResult => {
      const path = this.getRelativePath(fullFile)
      return suitesMap[path] ?? (suitesMap[path] = new TestSuiteResult(path, []))
    }

    const processPassingTests = (tests: MochawesomeJsonTest[], suiteName: string): void => {
      const passingTests = tests?.filter(test => test.pass)

      if (passingTests) {
        for (const passingTest of passingTests) {
          const suite = getSuite(suiteName)
          this.processTest(suite, passingTest, 'success')
        }
      }
    }

    const processFailingTests = (tests: MochawesomeJsonTest[], suiteName: string): void => {
      const failingTests = tests?.filter(test => test.fail)

      if (failingTests) {
        for (const failingTest of failingTests) {
          const suite = getSuite(suiteName)
          this.processTest(suite, failingTest, 'failed')
        }
      }
    }

    const processPendingTests = (tests: MochawesomeJsonTest[], suiteName: string): void => {
      const pendingTests = tests?.filter(test => test.skipped)

      if (pendingTests) {
        for (const pendingTest of pendingTests) {
          const suite = getSuite(suiteName)
          this.processTest(suite, pendingTest, 'skipped')
        }
      }
    }

    const processAllTests = (tests: MochawesomeJsonTest[], suiteName: string): void => {
      processPassingTests(tests, suiteName)
      processFailingTests(tests, suiteName)
      processPendingTests(tests, suiteName)
    }

    // Handle nested suites
    const processNestedSuites = (suite: MochawesomeJsonSuite, nestedSuiteIndex: number, suiteName: string): void => {
      // Process suite tests
      processAllTests(suite.tests, suiteName)

      for (const innerSuite of suite.suites) {
        // Process inner suite tests
        processAllTests(innerSuite.tests, suiteName)

        if (innerSuite?.suites[nestedSuiteIndex]?.suites.length > 0) {
          processNestedSuites(innerSuite, 0, suiteName)
        } else {
          processAllTests(innerSuite?.suites[nestedSuiteIndex]?.tests, suiteName)
          nestedSuiteIndex++

          // TODO - Figure out how to get 1.1.1.1.2 - suites with more than one object in them
        }
      }
    }

    // Process Mochawesome Data
    for (const result of results) {
      const suites = result?.suites
      const filePath = result?.fullFile
      const suitelessTests = result?.tests

      // Process tests that aren't in a suite
      if (suitelessTests?.length > 0) {
        processAllTests(suitelessTests, filePath)
      }

      // Process tests that are in a suite
      if (suites?.length > 0) {
        for (const suite of suites) {
          processNestedSuites(suite, 0, filePath ? filePath : suite.title)
        }
      }
    }

    const mappedSuites = Object.values(suitesMap)

    return new TestRunResult(resultsPath, mappedSuites, mochawesome.stats.duration)
  }

  private processTest(suite: TestSuiteResult, test: MochawesomeJsonTest, result: TestExecutionResult): void {
    const groupName =
      test.fullTitle !== test.title
        ? test.fullTitle.substr(0, test.fullTitle.length - test.title.length).trimEnd()
        : null

    let group = suite.groups.find(grp => grp.name === groupName)
    if (group === undefined) {
      group = new TestGroupResult(groupName, [])
      suite.groups.push(group)
    }

    const error = this.getTestCaseError(test)
    const testCase = new TestCaseResult(test.title, result, test.duration ?? 0, error)
    group.tests.push(testCase)
  }

  private getTestCaseError(test: MochawesomeJsonTest): TestCaseError | undefined {
    const details = test.err.estack
    const message = test.err.message
    if (details === undefined) {
      return undefined
    }

    let path
    let line

    const src = getExceptionSource(details, this.options.trackedFiles, file => this.getRelativePath(file))
    if (src) {
      path = src.path
      line = src.line
    }

    return {
      path,
      line,
      message,
      details
    }
  }

  private getRelativePath(path: string): string {
    path = normalizeFilePath(path)
    const workDir = this.getWorkDir(path)
    if (workDir !== undefined && path.startsWith(workDir)) {
      path = path.substr(workDir.length)
    }
    return path
  }

  private getWorkDir(path: string): string | undefined {
    return (
      this.options.workDir ??
      this.assumedWorkDir ??
      (this.assumedWorkDir = getBasePath(path, this.options.trackedFiles))
    )
  }
}
