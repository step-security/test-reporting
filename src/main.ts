import * as core from '@actions/core'
import * as github from '@actions/github'
import {createHash} from 'crypto'
import {GitHub} from '@actions/github/lib/utils'
import axios, {isAxiosError} from 'axios'

import {ArtifactProvider} from './input-providers/artifact-provider'
import {LocalFileProvider} from './input-providers/local-file-provider'
import {FileContent} from './input-providers/input-provider'
import {ParseOptions, TestParser} from './test-parser'
import {TestRunResult} from './test-results'
import {getAnnotations} from './report/get-annotations'
import {getReport} from './report/get-report'

import {DartJsonParser} from './parsers/dart-json/dart-json-parser'
import {DotnetTrxParser} from './parsers/dotnet-trx/dotnet-trx-parser'
import {JavaJunitParser} from './parsers/java-junit/java-junit-parser'
import {JestJunitParser} from './parsers/jest-junit/jest-junit-parser'
import {MochaJsonParser} from './parsers/mocha-json/mocha-json-parser'
import {MochawesomeJsonParser} from './parsers/mochawesome-json/mochawesome-json-parser'

import {normalizeDirPath, normalizeFilePath} from './utils/path-utils'
import {getCheckRunContext} from './utils/github-utils'
import {Outputs} from './utils/constants'

async function validateSubscription(): Promise<void> {
  const API_URL = `https://agent.api.stepsecurity.io/v1/github/${process.env.GITHUB_REPOSITORY}/actions/subscription`

  try {
    await axios.get(API_URL, {timeout: 3000})
  } catch (error) {
    if (isAxiosError(error) && error.response?.status === 403) {
      core.error('Subscription is not valid. Reach out to support@stepsecurity.io')
      process.exit(1)
    } else {
      core.info('Timeout or API not reachable. Continuing to next step.')
    }
  }
}

async function main(): Promise<void> {
  try {
    await validateSubscription()
    const testReporter = new TestReporter()
    await testReporter.run()
  } catch (error: any) {
    core.setFailed(error.message)
  }
}

function createSlugPrefix(): string {
  const step_summary = process.env['GITHUB_STEP_SUMMARY']
  if (!step_summary || step_summary === '') {
    return ''
  }
  const hash = createHash('sha1')
  hash.update(step_summary)
  return hash.digest('hex').substring(0, 8)
}

class TestReporter {
  readonly artifact = core.getInput('artifact', {required: false})
  readonly name = core.getInput('name', {required: true})
  readonly path = core.getInput('path', {required: true})
  readonly pathReplaceBackslashes = core.getInput('path-replace-backslashes', {required: false}) === 'true'
  readonly reporter = core.getInput('reporter', {required: true})
  readonly listSuites = core.getInput('list-suites', {required: true}) as 'all' | 'failed'
  readonly listTests = core.getInput('list-tests', {required: true}) as 'all' | 'failed' | 'none'
  readonly maxAnnotations = parseInt(core.getInput('max-annotations', {required: true}))
  readonly failOnError = core.getInput('fail-on-error', {required: true}) === 'true'
  readonly workDirInput = core.getInput('working-directory', {required: false})
  readonly onlySummary = core.getInput('only-summary', {required: false}) === 'true'
  readonly outputTo = core.getInput('output-to', {required: false})
  readonly token = core.getInput('token', {required: true})
  readonly slugPrefix: string = ''
  readonly octokit: InstanceType<typeof GitHub>
  readonly context = getCheckRunContext()

  constructor() {
    this.octokit = github.getOctokit(this.token)

    if (this.listSuites !== 'all' && this.listSuites !== 'failed') {
      core.setFailed(`Input parameter 'list-suites' has invalid value`)
      return
    }

    if (this.listTests !== 'all' && this.listTests !== 'failed' && this.listTests !== 'none') {
      core.setFailed(`Input parameter 'list-tests' has invalid value`)
      return
    }

    if (isNaN(this.maxAnnotations) || this.maxAnnotations < 0 || this.maxAnnotations > 50) {
      core.setFailed(`Input parameter 'max-annotations' has invalid value`)
      return
    }

    if (this.outputTo !== 'checks' && this.outputTo !== 'step-summary') {
      core.setFailed(`Input parameter 'output-to' has invalid value`)
      return
    }

    if (this.outputTo === 'step-summary') {
      this.slugPrefix = createSlugPrefix()
    }
  }

  async run(): Promise<void> {
    if (this.workDirInput) {
      core.info(`Changing directory to '${this.workDirInput}'`)
      process.chdir(this.workDirInput)
    }

    core.info(`Check runs will be created with SHA=${this.context.sha}`)

    // Split path pattern by ',' and optionally convert all backslashes to forward slashes
    // fast-glob (micromatch) always interprets backslashes as escape characters instead of directory separators
    const pathsList = this.path.split(',')
    const pattern = this.pathReplaceBackslashes ? pathsList.map(normalizeFilePath) : pathsList

    const inputProvider = this.artifact
      ? new ArtifactProvider(
          this.octokit,
          this.artifact,
          this.name,
          pattern,
          this.context.sha,
          this.context.runId,
          this.token
        )
      : new LocalFileProvider(this.name, pattern)

    const parseErrors = this.maxAnnotations > 0
    const trackedFiles = await inputProvider.listTrackedFiles()
    const workDir = this.artifact ? undefined : normalizeDirPath(process.cwd(), true)

    core.info(`Found ${trackedFiles.length} files tracked by GitHub`)

    const options: ParseOptions = {
      workDir,
      trackedFiles,
      parseErrors
    }

    core.info(`Using test report parser '${this.reporter}'`)
    const parser = this.getParser(this.reporter, options)

    const results: TestRunResult[] = []
    const input = await inputProvider.load()
    for (const [reportName, files] of Object.entries(input)) {
      try {
        core.startGroup(`Creating test report ${reportName}`)
        const tr = await this.createReport(parser, reportName, files)
        results.push(...tr)
      } finally {
        core.endGroup()
      }
    }

    const isFailed = results.some(tr => tr.result === 'failed')
    const conclusion = isFailed ? 'failure' : 'success'
    const passed = results.reduce((sum, tr) => sum + tr.passed, 0)
    const failed = results.reduce((sum, tr) => sum + tr.failed, 0)
    const skipped = results.reduce((sum, tr) => sum + tr.skipped, 0)
    const time = results.reduce((sum, tr) => sum + tr.time, 0)

    core.setOutput('conclusion', conclusion)
    core.setOutput('passed', passed)
    core.setOutput('failed', failed)
    core.setOutput('skipped', skipped)
    core.setOutput('time', time)

    if (this.failOnError && isFailed) {
      core.setFailed(`Failed tests were found and 'fail-on-error' option is set to ${this.failOnError}`)
      return
    }

    if (results.length === 0) {
      core.setFailed(`No test report files were found`)
      return
    }
  }

  async createReport(parser: TestParser, name: string, files: FileContent[]): Promise<TestRunResult[]> {
    if (files.length === 0) {
      core.warning(`No file matches path ${this.path}`)
      return []
    }

    const results: TestRunResult[] = []
    for (const {file, content} of files) {
      core.info(`Processing test results from ${file}`)
      const tr = await parser.parse(file, content)
      results.push(tr)
    }

    let createResp = null
    let baseUrl = ''
    let check_run_id = 0

    switch (this.outputTo) {
      case 'checks': {
        core.info(`Creating check run ${name}`)
        createResp = await this.octokit.rest.checks.create({
          head_sha: this.context.sha,
          name,
          status: 'in_progress',
          output: {
            title: name,
            summary: ''
          },
          ...github.context.repo
        })
        baseUrl = createResp.data.html_url as string
        check_run_id = createResp.data.id
        break
      }
      case 'step-summary': {
        const run_attempt = process.env['GITHUB_RUN_ATTEMPT'] ?? 1
        baseUrl = `https://github.com/${github.context.repo.owner}/${github.context.repo.repo}/actions/runs/${github.context.runId}/attempts/${run_attempt}`
        break
      }
    }

    core.info('Creating report summary')
    const {listSuites, listTests, onlySummary, slugPrefix} = this
    const summary = getReport(results, {listSuites, listTests, baseUrl, slugPrefix, onlySummary})

    core.info('Creating annotations')
    const annotations = getAnnotations(results, this.maxAnnotations)

    const isFailed = results.some(tr => tr.result === 'failed')
    const conclusion = isFailed ? 'failure' : 'success'

    const passed = results.reduce((sum, tr) => sum + tr.passed, 0)
    const failed = results.reduce((sum, tr) => sum + tr.failed, 0)
    const skipped = results.reduce((sum, tr) => sum + tr.skipped, 0)
    const shortSummary = `${passed} passed, ${failed} failed and ${skipped} skipped `

    core.info(`Updating check run conclusion (${conclusion}) and output`)
    switch (this.outputTo) {
      case 'checks': {
        const resp = await this.octokit.rest.checks.update({
          check_run_id,
          conclusion,
          status: 'completed',
          output: {
            title: shortSummary,
            summary,
            annotations
          },
          ...github.context.repo
        })
        core.info(`Check run create response: ${resp.status}`)
        core.info(`Check run URL: ${resp.data.url}`)
        core.info(`Check run HTML: ${resp.data.html_url}`)
        core.setOutput(Outputs.runHtmlUrl, `${resp.data.html_url}`)
        break
      }
      case 'step-summary': {
        core.summary.addRaw(`# ${shortSummary}`)
        core.summary.addRaw(summary)
        await core.summary.write()
        for (const annotation of annotations) {
          let fn
          switch (annotation.annotation_level) {
            case 'failure':
              fn = core.error
              break
            case 'warning':
              fn = core.warning
              break
            case 'notice':
              fn = core.notice
              break
            default:
              continue
          }

          fn(annotation.message, {
            title: annotation.title,
            file: annotation.path,
            startLine: annotation.start_line,
            endLine: annotation.end_line,
            startColumn: annotation.start_column,
            endColumn: annotation.end_column
          })
        }
        break
      }
    }

    return results
  }

  getParser(reporter: string, options: ParseOptions): TestParser {
    switch (reporter) {
      case 'dart-json':
        return new DartJsonParser(options, 'dart')
      case 'dotnet-trx':
        return new DotnetTrxParser(options)
      case 'flutter-json':
        return new DartJsonParser(options, 'flutter')
      case 'java-junit':
        return new JavaJunitParser(options)
      case 'jest-junit':
        return new JestJunitParser(options)
      case 'mocha-json':
        return new MochaJsonParser(options)
      case 'mochawesome-json':
        return new MochawesomeJsonParser(options)
      default:
        throw new Error(`Input variable 'reporter' is set to invalid value '${reporter}'`)
    }
  }
}

main()
