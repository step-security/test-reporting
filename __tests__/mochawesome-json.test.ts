import * as fs from 'fs'
import * as path from 'path'

import {MochawesomeJsonParser} from '../src/parsers/mochawesome-json/mochawesome-json-parser'
import {ParseOptions} from '../src/test-parser'
import {getReport} from '../src/report/get-report'
import {normalizeFilePath} from '../src/utils/path-utils'

describe('mochawesome-json tests', () => {
  it('produces empty test run result when there are no test cases', async () => {
    const fixturePath = path.join(__dirname, 'fixtures', 'empty', 'mochawesome-json.json')
    const filePath = normalizeFilePath(path.relative(__dirname, fixturePath))
    const fileContent = fs.readFileSync(fixturePath, {encoding: 'utf8'})

    const opts: ParseOptions = {
      parseErrors: true,
      trackedFiles: []
    }

    const parser = new MochawesomeJsonParser(opts)
    const result = await parser.parse(filePath, fileContent)
    expect(result.tests).toBe(0)
    expect(result.result).toBe('success')
  })

  it('report from ./reports/mochawesome-json test results matches snapshot', async () => {
    const fixturePath = path.join(__dirname, 'fixtures', 'mochawesome-json.json')
    const outputPath = path.join(__dirname, '__outputs__', 'mochawesome-json.md')
    const filePath = normalizeFilePath(path.relative(__dirname, fixturePath))
    const fileContent = fs.readFileSync(fixturePath, {encoding: 'utf8'})

    const opts: ParseOptions = {
      parseErrors: true,
      trackedFiles: ['test/main.test.js', 'test/second.test.js', 'lib/main.js']
    }

    const parser = new MochawesomeJsonParser(opts)
    const result = await parser.parse(filePath, fileContent)
    expect(result).toMatchSnapshot()

    const report = getReport([result])
    fs.mkdirSync(path.dirname(outputPath), {recursive: true})
    fs.writeFileSync(outputPath, report)
  })

  // TODO - External
  // it('report from mochawesomejs/mochawesome test results matches snapshot', async () => {
  //   const fixturePath = path.join(__dirname, 'fixtures', 'external', 'mochawesome', 'mochawesome-test-results.json')
  //   const trackedFilesPath = path.join(__dirname, 'fixtures', 'external', 'mochawesome', 'files.txt')
  //   const outputPath = path.join(__dirname, '__outputs__', 'mochawesome-test-results.md')
  //   const filePath = normalizeFilePath(path.relative(__dirname, fixturePath))
  //   const fileContent = fs.readFileSync(fixturePath, {encoding: 'utf8'})

  //   const trackedFiles = fs.readFileSync(trackedFilesPath, {encoding: 'utf8'}).split(/\n\r?/g)
  //   const opts: ParseOptions = {
  //     parseErrors: true,
  //     trackedFiles
  //   }

  //   const parser = new MochawesomeJsonParser(opts)
  //   const result = await parser.parse(filePath, fileContent)
  //   expect(result).toMatchSnapshot()

  //   const report = getReport([result])
  //   fs.mkdirSync(path.dirname(outputPath), {recursive: true})
  //   fs.writeFileSync(outputPath, report)
  // })
})
