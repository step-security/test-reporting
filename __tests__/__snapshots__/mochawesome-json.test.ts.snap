// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mochawesome-json tests report from ./reports/mochawesome-json test results matches snapshot 1`] = `
TestRunResult {
  "path": "fixtures/mochawesome-json.json",
  "suites": [
    TestSuiteResult {
      "groups": [
        TestGroupResult {
          "name": null,
          "tests": [
            TestCaseResult {
              "error": {
                "details": "Error: Timeout of 1ms exceeded. For async tests and hooks, ensure \"done()\" is called; if returning a Promise, ensure it resolves. (/Users/work/Source/Repos/thirdparty/phoenix-test-reporter/reports/mochawesome/test/second.test.js)
    at listOnTimeout (node:internal/timers:557:17)
    at processTimers (node:internal/timers:500:7)",
                "line": undefined,
                "message": "Error: Timeout of 1ms exceeded. For async tests and hooks, ensure \"done()\" is called; if returning a Promise, ensure it resolves. (/Users/work/Source/Repos/thirdparty/phoenix-test-reporter/reports/mochawesome/test/second.test.js)",
                "path": undefined,
              },
              "name": "Timeout test",
              "result": "failed",
              "time": 3,
            },
          ],
        },
        TestGroupResult {
          "name": "Test 1",
          "tests": [
            TestCaseResult {
              "error": undefined,
              "name": "Passing test",
              "result": "success",
              "time": 0,
            },
          ],
        },
        TestGroupResult {
          "name": "Test 1 Test 1.1",
          "tests": [
            TestCaseResult {
              "error": {
                "details": "Error: Some error
    at Object.throwError (lib/main.js:2:9)
    at Context.<anonymous> (test/main.test.js:15:11)
    at processImmediate (node:internal/timers:464:21)",
                "line": 2,
                "message": "Error: Some error",
                "path": "lib/main.js",
              },
              "name": "Exception in target unit",
              "result": "failed",
              "time": 0,
            },
            TestCaseResult {
              "error": {
                "details": "AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:

false !== true

    at Context.<anonymous> (test/main.test.js:11:14)
    at processImmediate (node:internal/timers:464:21)",
                "line": 11,
                "message": "AssertionError: Expected values to be strictly equal:

false !== true
",
                "path": "test/main.test.js",
              },
              "name": "Failing test",
              "result": "failed",
              "time": 2,
            },
          ],
        },
        TestGroupResult {
          "name": "Test 2",
          "tests": [
            TestCaseResult {
              "error": {
                "details": "Error: Some error
    at Context.<anonymous> (test/main.test.js:22:11)
    at processImmediate (node:internal/timers:464:21)",
                "line": 22,
                "message": "Error: Some error",
                "path": "test/main.test.js",
              },
              "name": "Exception in test",
              "result": "failed",
              "time": 0,
            },
          ],
        },
      ],
      "name": "nofile",
      "totalTime": undefined,
    },
  ],
  "totalTime": 14,
}
`;
